cmd/ido2icingadb/convert.go

```diff
@@ -0,0 +1,893 @@
+				ObjectType:    typ,
+				HostId:        hostId,
+				ServiceId:     serviceId,
+			},
+			DowntimeStart:    startTime,
+			HasBeenCancelled: icingadbTypes.Bool{Bool: row.WasCancelled != 0, Valid: true},
+			CancelTime:       cancelTime,
+			EndTime:          endTime,
+		}
+
+		s.DowntimeEnd.History = s
+		sla = append(sla, s)
+
+		checkpoint = row.DowntimehistoryId
+	}
+
+	icingaDbInserts = [][]interface{}{downtimeHistory, allHistory, sla}
+	return
+}
+
+// FlappingEnd updates an already migrated start event with the end event info.
+type FlappingEnd struct {
+	Id      icingadbTypes.Binary    `json:"id"`
+	EndTime icingadbTypes.UnixMilli `json:"end_time"`
+}
+
+// Assert interface compliance.
+var _ contracts.TableNamer = (*FlappingEnd)(nil)
+
+// TableName implements the contracts.TableNamer interface.
+func (*FlappingEnd) TableName() string {
+	return "flapping_history"
+}
+
+type flappingRow = struct {
+	FlappinghistoryId  uint64
+	EventTime          int64
+	EventTimeUsec      uint32
+	EventType          uint16
+	PercentStateChange sql.NullFloat64
+	LowThreshold       float64
+	HighThreshold      float64
+	ObjecttypeId       uint8
+	Name1              string
+	Name2              string
+}
+
+func convertFlappingRows(
+	env string, envId, endpointId icingadbTypes.Binary,
+	selectCache func(dest interface{}, query string, args ...interface{}), _ *sqlx.Tx, idoRows []flappingRow,
+) (icingaDbUpdates, icingaDbInserts [][]interface{}, checkpoint interface{}) {
+	if len(idoRows) < 1 {
+		return
+	}
+
+	var cached []struct {
+		HistoryId     uint64
+		EventTime     int64
+		EventTimeUsec uint32
+	}
+	selectCache(
+		&cached, "SELECT history_id, event_time, event_time_usec FROM end_start_time WHERE history_id BETWEEN ? AND ?",
+		idoRows[0].FlappinghistoryId, idoRows[len(idoRows)-1].FlappinghistoryId,
+	)
+
+	// Needed for start time (see below).
+	cachedById := make(map[uint64]icingadbTypes.UnixMilli, len(cached))
+	for _, c := range cached {
+		cachedById[c.HistoryId] = convertTime(c.EventTime, c.EventTimeUsec)
+	}
+
+	var flappingHistory, flappingHistoryUpdates, allHistory []interface{}
+	for _, row := range idoRows {
+		ts := convertTime(row.EventTime, row.EventTimeUsec)
+
+		// Needed for ID (see below).
+		var start icingadbTypes.UnixMilli
+		if row.EventType == 1001 { // end
+			var ok bool
+			start, ok = cachedById[row.FlappinghistoryId]
+
+			if !ok {
+				continue
+			}
+		} else {
+			start = ts
+		}
+
+		name := row.Name1
+		if row.Name2 != "" {
+			name += "!" + row.Name2
+		}
+
+		typ := objectTypes[row.ObjecttypeId]
+		hostId := calcObjectId(env, row.Name1)
+		serviceId := calcServiceId(env, row.Name1, row.Name2)
+		startTime := float64(start.Time().UnixMilli())
+		flappingHistoryId := hashAny([]interface{}{env, name, startTime})
+
+		if row.EventType == 1001 { // end
+			// The start counterpart should already have been inserted.
+			flappingHistoryUpdates = append(flappingHistoryUpdates, &FlappingEnd{flappingHistoryId, ts})
```

Comment:
Should also set `percent_state_change_end` to the value of `percent_state_change`.